core/src/main/java/org/pentaho/di/core/xml/XMLFormatter.java

```diff
@@ -0,0 +1,245 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2016-2017 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+package org.pentaho.di.core.xml;
+
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import javax.xml.stream.XMLStreamWriter;
+
+/**
+ * XML formatting for better VCS diff.
+ * 
+ * It preserve formatting only in cases: 1) inside one tag if there are only characters, 2) in comments, 3) if there are
+ * some characters outside tags
+ *
+ * @author Alexander Buloichik
+ */
+public class XMLFormatter {
+  private static final String STEP_PREFIX = "  ";
+
+  private static XMLInputFactory INPUT_FACTORY = XMLInputFactory.newInstance();
+  private static XMLOutputFactory OUTPUT_FACTORY = XMLOutputFactory.newInstance();
+
+  static {
+    INPUT_FACTORY.setProperty( XMLInputFactory.IS_COALESCING, false );
+  }
+
+  public static String format( String xml ) {
+    XMLStreamReader rd = null;
+    XMLStreamWriter wr = null;
+
+    StringWriter result = new StringWriter();
+    try {
+      rd = INPUT_FACTORY.createXMLStreamReader( new StringReader( xml ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Parsing XML input using a weakly configured XML parser can lead to XML external entity (XXE) injection. To protect against XXE attacks, be sure to securely configure your XML parser.

[Learn more](https://skb.highcastle.a2z.com/implementations/189)